infinispan-remote/src/main/java/org/hibernate/ogm/datastore/infinispanremote/InfinispanRemoteDialect.java

```diff
@@ -0,0 +1,380 @@
+					firstIteration = false;
+				}
+				else {
+					bqEnd = bqEnd.and().having( fieldName ).eq( key.getColumnValues()[i] );
+				}
+			}
+			Query query = bqEnd.toBuilder().build();
+			Map<RowKey,Map<String, Object>> resultsCollector = new HashMap<>();
+			try ( CloseableIterator<Entry<Object,Object>> iterator = c.retrieveEntriesByQuery( query, null, 100 ) ) {
+				while ( iterator.hasNext() ) {
+					Entry<Object,Object> e  = iterator.next();
+					ProtostreamPayload value = ( (ProtostreamPayload) e.getValue() );
+					Map<String, Object> entryObject = value.toMap();
+					RowKey entryKey = value.asRowKey( key );
+					resultsCollector.put( entryKey, entryObject );
+				}
+			}
+			return resultsCollector;
+		} );
+	}
+
+	@Override
+	public Association createAssociation(AssociationKey key, AssociationContext associationContext) {
+		Map<RowKey, Map<String, Object>> associationMap = new HashMap<RowKey, Map<String,Object>>();
+		return new Association( new MapAssociationSnapshot( associationMap ) );
+	}
+
+	@Override
+	public void insertOrUpdateAssociation(AssociationKey key, Association association, AssociationContext associationContext) {
+		final String cacheName = key.getTable();
+		final ProtoStreamMappingAdapter mapper = provider.getDataMapperForCache( cacheName );
+		final boolean embeddedInEntity = associatonTypeIsForeignKeyEmbeddedInEntity( key, association, mapper );
+		if ( embeddedInEntity ) {
+			insertOrUpdateAssociationEmbeddedInEntity( key, association, associationContext, mapper, cacheName );
+		}
+		else {
+			insertOrUpdateAssociationMappedAsDedicatedEntries( key, association, associationContext, mapper, cacheName );
+		}
+	}
+
+	private void insertOrUpdateAssociationMappedAsDedicatedEntries(AssociationKey key, Association association, AssociationContext associationContext,
+			ProtoStreamMappingAdapter mapper, String cacheName) {
+		log.debugf( "insertOrUpdateAssociation for key '%s' on cache '%s', mapped as dedicated entries in ad-hoc table", key, cacheName );
+		final List<AssociationOperation> operations = association.getOperations();
+		for ( AssociationOperation ao : operations ) {
+			AssociationOperationType type = ao.getType();
+			RowKey rowKey = ao.getKey();
+			ProtostreamId idBuffer = mapper.createIdPayload( rowKey.getColumnNames(), rowKey.getColumnValues() );
+			switch ( type ) {
+				case PUT:
+					ProtostreamPayload valuePayloadForPut = mapper.createValuePayload( ao.getValue() );
+					mapper.withinCacheEncodingContext( c -> c.put( idBuffer, valuePayloadForPut ) );
+					break;
+				case REMOVE:
+					mapper.withinCacheEncodingContext( c -> c.remove( idBuffer ) );
+					break;
+				case CLEAR:
+					throw new AssertionFailure( "Request for CLEAR operation on an association mapped to dedicated entries. Makes no sense?" );
+			}
+		}
+	}
+
+	private void insertOrUpdateAssociationEmbeddedInEntity(AssociationKey key, Association association, AssociationContext associationContext,
```

Comment:
In this case if I understand correctly, you literally store all of the associaton columns (fk?) flat in the object.

```
User {
   String name;
   String addressId1;
   String addressId2;
}
// assuming Address is a composite key of id1 and id2
```

Correct?

In mongodb, we do create an intermediary object, something like

```
```
User {
   String name;
   AddressId address {
       String addressId1;
       String addressId2;
   }
}
// assuming Address is a composite key of id1 and id2
```

PS @gunnarmorling, I am correct on MongoDB right?